### PR TITLE
tools: Refactor `--with` logic into composer-update-monorepo.sh

### PR DESCRIPTION
**Pull request by anomiex**
#### Changes proposed in this Pull Request:

In #25989 we added logic to version-packages.sh to use `--with` when calling composer-update-monorepo.sh to force monorepo deps in the "^1.2.3" style to install version 1.2.3 exactly, rather than picking up any later version, to better work in release branches where that style of dep is used.

And in #26072, we did the same thing for check-intra-monorepo-deps.sh for the same reason.

Chances are any other caller of composer-update-monorepo.sh is going to want the same too, so let's just move the logic there.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

p1662751617679859/1662483262.626649-slack-C01U2KGS2PQ

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Run version-packages.sh, as if to create a Jetpack release. See that things get updated as expected.
* Edit `projects/packages/videopress/CHANGELOG.md` to reflect a newer version than has actually been released. Run version-packages.sh, it should fail.
* Edit `projects/packages/videopress/CHANGELOG.md` to remove the newest changelog entry. Run version-packages.sh, it should succeed and should lock the videopress package to the older version in composer.lock.
* Check out c405506d5b6d1e4951d99aa188b918b4bb0f1448, cherry-pick the relevant changes (`git cherry-pick 7859236 ba6d621 f84d97c d2ed119`), then run `COMPOSER_ROOT_VERSION=dev-trunk tools/changelogger-release.sh packages/publicize -p`.
  * [ ] The dep on my-jetpack in `projects/plugins/jetpack/composer.json` should be changed to "^2.0.3".
  * [ ] Many other deps should change from "^1.2" to "^1.2.3" style.
  * [ ] The change in `projects/plugins/jetpack/composer.lock` is only to the record for the publicize package. No other packages change.

### CODE REVIEW

**Comment by sdixon194** [comment]
Made it as far as:

> Check out https://github.com/Automattic/jetpack/commit/c405506d5b6d1e4951d99aa188b918b4bb0f1448, cherry-pick the relevant changes (git cherry-pick 7859236 ba6d621 f84d97c d2ed119), then run COMPOSER_ROOT_VERSION=dev-trunk tools/changelogger-release.sh packages/publicize -p.

It doesn't look like git can find those commits to cherry pick (and the first one doesn't look like a commit hash?)

**Comment by anomiex** [comment]
> It doesn't look like git can find those commits to cherry pick (and the first one doesn't look like a commit hash?)

Ah, you might have to fetch them specifically since they're from a past PR. `git fetch origin f84d97c6b3c9300a5899750e51a20e851a57f4b0` should do it, since the other two are ancestors of that one.

File: tools/composer-update-monorepo.sh [hunk 1]
```diff
@@ -77,16 +77,26 @@ if [[ -z "$DIR" ]]; then
...
 		if $ROOT_REQS; then
 			composer info --working-dir="$DIR" --locked --direct --name-only | sed -e 's/ *$//'
 		fi
 	) | sort -u
 )
-COMPOSER_ROOT_VERSION=dev-trunk composer update "${COMPOSER_ARGS[@]}" --working-dir="$DIR" -- "${TO_UPDATE[@]}"
+
+# Any deps that look like "^1.2.3" should be locked at that version, as it's probably a release branch.
+WITH=()
+TMP="$(jq -r --arg packages $'\n'"$PACKAGES"$'\n' '.require // {}, .["require-dev"] // {} | to_entries[] | select( ( "\n\( .key )\n" | inside( $packages ) ) and ( .value | test( "^\\^[0-9]+\\.[0-9]+\\.[0-9]+$" ) ) ) | "--with=\( .key )=\( .value[1:] )"' "$DIR/composer.json")"
```
Comment (sdixon194) [inline tools/composer-update-monorepo.sh]: This will only work for versions like `1.2.3`, but not `1.2.3.4`

Comment (anomiex) [reply tools/composer-update-monorepo.sh]: Fixed!

**Review by sdixon194 — APPROVED** [review]

**Pull request merged**